pcdsdevices/pneumatic.py

```diff
@@ -45,27 +43,32 @@ class BeckhoffPneumatic(BaseInterface, LightpathMixin):
     error_message = Cpt(EpicsSignalRO, ':PLC:sErrorMessage')
     position_state = Cpt(EpicsSignalRO, ':nPositionState', kind='hinted')
 
-    def insert(self, time_out=30):
+    def insert(self, wait: bool = False, timeout: float = 10.0) -> Status:
         """
         Method for inserting Beckhoff Pneumatic Actuator
+        default times out at 10 seconds, setting wait to true
+        will block indefinietly until the PLC marks not busy and
+        done or errors.
         """
+        status = Status(timeout=timeout) if not wait else Status()
+
         if self.insert_ok.get():
             self.insert_signal.put(1)
-            start_time = time.time()
-            time_elapsed = 0
-            while (time_elapsed < time_out):
-                time_elapsed = time.time() - start_time
-                if not self.done.get() and self.busy.get():
-                    continue
-                else:
-                    break
-            if self.error.get():
-                raise RuntimeError(self.error_message.get())
-            if time_elapsed >= time_out:
-                raise RuntimeError("Insertion Failed Timed Out")
-
+            # three ways to leave this loop
+            # 1 true done and not busy
+            # 2 error
+            # 3 timeout
+            while (not self.done.get() and self.busy.get()):
```

Comment:
If I understand status correctly, one instantiated, it will keep its own count and raise an exception if it expires.

Reply:
Yeah, that's correct. `Status` does a lot of the timeout stuff for you and has helpful methods like `status.wait`. It can also be used to communicate error status without blocking, since you can make a background thread or a subscription callback check for errors and apply them to the `Status` object outside of the main thread.

A common pattern is:
- subscribe to self.done with a function that marks `Status` as finished or sets the exception, then unsubscribes itself
- optionally wait on the status object
- return the status object

Another common pattern is:
- create a thread the waits for done in a loop with some sleeps thrown in, then marks `Status` as finished or sets the exception
- optionally wait on the status object
- return the status object

All of this isn't needed to just make the pneumatic work but it's the most complete way in ophyd to report error states. Otherwise, our main asynchronous error reporting tool (for the `wait=False` case) would be logging and print statements which aren't super useful for responding to errors programatically.

Reply:
Added a sketch of what I think the first pattern should look like. Not sure how I access the PV value in the call back subscription.